### PR TITLE
Initial version of dynamichmc_inference

### PR DESCRIPTION
**Pull request by Vaibhavdixit02**
Initial version of dynamichmc_inference which serves as the wrapper for DynamicHMC.jl for using it as backend in DiffEqBayes.jl

### CODE REVIEW

File: src/dynamichmc_inference.jl [hunk 1]
```diff
@@ -0,0 +1,47 @@
+struct LotkaVolterraPosterior{Problem, Data, A_Prior, ObservationTimes, ErrorDist}
```
Comment (ChrisRackauckas) [inline src/dynamichmc_inference.jl]: This naming is still tied to the model. Name it after DynamicHMC

File: src/dynamichmc_inference.jl [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    ℓ + logpdf(a_prior, a)
+end
+
+function dynamichmc_inference(prob1::DEProblem,data,priors,t,σ = 0.01)
+    P = LotkaVolterraPosterior(prob1, data, priors, t, Normal(0.0, σ))
+
+    parameter_transformation = TransformationTuple((bridge(ℝ, ℝ⁺, ))) # assuming a > 0
+
+    PT = TransformLogLikelihood(P, parameter_transformation)
+    PTG = ForwardGradientWrapper(PT, zeros(1));
```
Comment (ChrisRackauckas) [inline src/dynamichmc_inference.jl]: Why 1?

File: src/dynamichmc_inference.jl [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    if !isfinite(ℓ) && (ℓ ≠ -Inf)
+        ℓ = -Inf                # protect against NaN etc, is it needed?
+    end
+    ℓ + logpdf(a_prior, a)
+end
+
+function dynamichmc_inference(prob1::DEProblem,data,priors,t,σ = 0.01)
+    P = LotkaVolterraPosterior(prob1, data, priors, t, Normal(0.0, σ))
+
+    parameter_transformation = TransformationTuple((bridge(ℝ, ℝ⁺, ))) # assuming a > 0
```
Comment (ChrisRackauckas) [inline src/dynamichmc_inference.jl]: This parameter transformation is model-dependent.

File: src/dynamichmc_inference.jl [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    PT = TransformLogLikelihood(P, parameter_transformation)
+    PTG = ForwardGradientWrapper(PT, zeros(1));
+
+# NOTE: starting from correct parameter is important, otherwise stepsize
+# adaptation is not handled well. would probably maximize PT in a real-life
+# setting.
+    PO = OnceDifferentiable(x -> -P(x), [2.0])
```
Comment (ChrisRackauckas) [inline src/dynamichmc_inference.jl]: Why this setup?

File: src/dynamichmc_inference.jl [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    a₀ = Optim.minimizer(optimize(a -> -P([a]), 0, 10))
```
Comment (ChrisRackauckas) [inline src/dynamichmc_inference.jl]: This optimizer only works for 1 parameter

File: src/dynamichmc_inference.jl [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    data::Data
+    a_prior::A_Prior
+    t::ObservationTimes
+    ϵ_dist::ErrorDist
+end
+
+function (P::LotkaVolterraPosterior)(θ)
+    @unpack problem, data, a_prior, t, ϵ_dist = P
+    a = θ[1]
+    try
```
Comment (ChrisRackauckas) [inline src/dynamichmc_inference.jl]: there's a better way than a try/catch, but let's get to that later

File: src/dynamichmc_inference.jl [hunk 7]
```diff
@@ -23,19 +23,18 @@ function (P::LotkaVolterraPosterior)(θ)
...
 
-# NOTE: starting from correct parameter is important, otherwise stepsize
-# adaptation is not handled well. would probably maximize PT in a real-life
-# setting.
-    PO = OnceDifferentiable(x -> -P(x), [2.0])
-    a₀ = Optim.minimizer(optimize(a -> -P([a]), 0, 10))
+    # NOTE: starting from correct parameter is important, otherwise stepsize
+    # adaptation is not handled well. would probably maximize PT in a real-life
+    # setting
+    a₀ = Optim.minimizer(optimize(a -> -P([a]), P.problem.tspan[1], P.problem.tspan[2]))
```
Comment (ChrisRackauckas) [inline src/dynamichmc_inference.jl]: Brent only works for a single parameter though.

**Comment by Vaibhavdixit02** [comment]
No this is just the single parameter version right now, wanted to get that
working first.

File: .travis.yml [hunk 8]
```diff
@@ -19,6 +19,9 @@ script:
   - export CMDSTAN_HOME=$HOME/cmdstan
   - cd $TRAVIS_BUILD_DIR
   - julia -e 'Pkg.clone(pwd()); Pkg.build("DiffEqBayes"); Pkg.test("DiffEqBayes"; coverage=true)'
+  - julia -e 'Pkg.clone("https://github.com/tpapp/ContinuousTransformations.jl.git"); Pkg.build("ContinuousTransformations")'
+  - julia -e 'Pkg.clone("https://github.com/tpapp/DiffWrappers.jl.git"); Pkg.build("DiffWrappers")'
```
[the remark this answers is not in the document]
Comment (tpapp) [reply .travis.yml]: I was waiting for 0.7 (named tuples, small unions) for easier interface and some optimizations & benchmarking, but can do it sooner if you prefer.

**Comment by tpapp** [comment]
I am in two minds about this: first, I am happy that people are using my package, second, what I sketched was a proof of concept, and I am not sure how robust it is. In any case, I will respond quickly if you need extensions/bugfixes to any of my libraries.

In any case, I will tag releases and register `DynamicHMC.jl` and `ContinuousTransformations.jl` in the upcoming weeks. This will provide a good opportunity to review the API (which, formally, is WIP ATM and can literally change at any point). Some further API redesign may happen at some point, so when I tag a version, please make sure you constrain the dependency version upper limit accordingly.

**Comment by ChrisRackauckas** [comment]
No worries. Honestly, it's fine if it's still changing: we just  won't document it until it's ready. But I thought we might as well get the proof of concept solidified into working code to keep updating, and start benchmarking to see how different strategies compare.

**Pull request merged**